engine/proofview.ml

```diff
@@ -700,7 +704,7 @@ let guard_no_unifiable =
 let unshelve l p =
   (* advance the goals in case of clear *)
   let l = undefined p.solution l in
-  { p with comb = p.comb@l }
+  { p with comb = p.comb@l; shelf = List.filter (fun g -> not (List.mem g l)) p.shelf }
```

Comment:
This seems a fix unrelated to this PR, am I right?

Reply:
It could be done before indeed.

Reply:
@gares @maximedenes Is just a separate commit enough or you want to see it in another PR?